File: src/comicengine/library.py
"""Story Library — catalog of UI stories + Reader / Only-Image editions."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from comicengine.config import OUTPUTS, ROOT
from comicengine.curation import curation
from comicengine.stories import list_stories

LIBRARY_DIR = OUTPUTS / "library"
CATALOG_PATH = LIBRARY_DIR / "catalog.json"


def _load_manifest(path: Path) -> dict[str, Any]:
    if not path.is_file():
        return {}
    try:
        data = json.loads(path.read_text())
        return data if isinstance(data, dict) else {}
    except (OSError, json.JSONDecodeError):
        return {}


def rebuild_catalog() -> dict[str, Any]:
    """Scan stories + Phase 7 / 7.5 manifests into outputs/library/catalog.json."""
    LIBRARY_DIR.mkdir(parents=True, exist_ok=True)
    entries: list[dict[str, Any]] = []
    for story in list_stories():
        sid = story["id"]
        composed = _load_manifest(OUTPUTS / "phase7" / sid / "manifest.json")
        image_only = _load_manifest(OUTPUTS / "phase7.5" / sid / "manifest.json")
        cur_items = curation.list(story_id=sid)
        cur_summary = {
            "approved": sum(1 for i in cur_items if i["status"] == "approved"),
            "rejected": sum(1 for i in cur_items if i["status"] == "rejected"),
            "pending": sum(1 for i in cur_items if i["status"] == "pending"),
            "regenerated": sum(1 for i in cur_items if i["status"] in {"regenerated", "regenerating"}),
            "total": len(cur_items),
            "panels": [
                {
                    "index": i.get("panel_index"),
                    "status": i.get("status"),
                    "note": i.get("note") or "",
                    "rating": i.get("rating"),
                    "suggestions": i.get("suggestions") or "",
                    "updated_at": i.get("updated_at"),
                }
                for i in cur_items
                if i.get("kind") == "panel"
            ],
            "avg_rating": (
                round(
                    sum(i.get("rating") or 0 for i in cur_items if i.get("kind") == "panel" and i.get("rating"))
                    / max(
                        sum(1 for i in cur_items if i.get("kind") == "panel" and i.get("rating")),
                        1,
                    ),
                    2,
                )
                if any(i.get("kind") == "panel" and i.get("rating") for i in cur_items)
                else None
            ),
            "episode_status": next(
                (i.get("status") for i in cur_items if i.get("kind") == "episode"),
                "pending",
            ),
        }
        entries.append(
            {
                "id": sid,
                "title": story.get("title") or sid,
                "topic": story.get("topic") or "",
                "voice": story.get("voice") or "",
                "panel_count": story.get("panel_count") or 0,
                "phase": story.get("phase") or "",
                "path": story.get("path") or "",
                "reader_href": story.get("href") or f"/stories/{sid}",
                "json_href": story.get("json_href") or f"/api/stories/{sid}",
                "disclaimer": story.get("disclaimer") or "",
                "curation": cur_summary,
                "editions": {
                    "reader": {
                        "label": "Reader (bubbles + captions)",
                        "phase": "phase7",
                        "available": bool(composed.get("webtoon_href") or composed.get("pdf_href")),
                        "webtoon_href": composed.get("webtoon_href")
                        or (f"/media/phase7/{sid}/webtoon.png" if composed else None),
                        "pdf_href": composed.get("pdf_href")
                        or (f"/media/phase7/{sid}/episode.pdf" if composed else None),
                        "webtoon_path": composed.get("webtoon_path"),
                        "pdf_path": composed.get("pdf_path"),
                        "assembled_panels": composed.get("assembled_panels"),
                    },
                    "image_only": {
                        "label": "Only Image (no bubbles)",
                        "phase": "phase7.5",
                        "available": bool(image_only.get("webtoon_href") or image_only.get("pdf_href")),
                        "webtoon_href": image_only.get("webtoon_href")
                        or (
                            f"/media/phase7.5/{sid}/webtoon_image_only.png" if image_only else None
                        ),
                        "pdf_href": image_only.get("pdf_href")
                        or (
                            f"/media/phase7.5/{sid}/episode_image_only.pdf" if image_only else None
                        ),
                        "webtoon_path": image_only.get("webtoon_path"),
                        "pdf_path": image_only.get("pdf_path"),
                        "assembled_panels": image_only.get("assembled_panels"),
                    },
                },
            }
        )

    catalog = {
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "count": len(entries),
        "stories": entries,
        "curation_summary": curation.summary(),
        "store_path": str(CATALOG_PATH.relative_to(ROOT)),
        "notes": (
            "Story Library stores dashboard editions: Reader (Phase 7 composed) "
            "and Only Image (Phase 7.5 raw panels), plus Phase 8/8.5 curation "
            "(status, rating, suggestions)."
        ),
    }
    CATALOG_PATH.write_text(json.dumps(catalog, indent=2))
    return catalog


def load_catalog(*, refresh: bool = False) -> dict[str, Any]:
    if refresh or not CATALOG_PATH.is_file():
        return rebuild_catalog()
    try:
        data = json.loads(CATALOG_PATH.read_text())
        if isinstance(data, dict) and "stories" in data:
            return data
    except (OSError, json.JSONDecodeError):
        pass
    return rebuild_catalog()

File: outputs/phase10/site/library.js
const $ = (id) => document.getElementById(id);

function esc(s) {
  return String(s ?? "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

function editionBlock(title, ed) {
  if (!ed) return "";
  const avail = ed.available;
  return `
    <div class="edition-card ${avail ? "" : "dim"}">
      <h3>${esc(title)}</h3>
      <p class="muted small">${esc(ed.label || "")} · ${esc(ed.phase || "")}</p>
      <div class="link-row">
        ${ed.webtoon_href && avail ? `<a class="btn" href="${esc(ed.webtoon_href)}" target="_blank" rel="noopener">Webtoon</a>` : ""}
        ${ed.pdf_href && avail ? `<a class="btn" href="${esc(ed.pdf_href)}" target="_blank" rel="noopener">PDF</a>` : ""}
        ${!avail ? `<span class="muted small">Not assembled</span>` : ""}
      </div>
      ${ed.webtoon_href && avail
        ? `<img class="webtoon-preview" src="${esc(ed.webtoon_href)}" alt="${esc(title)}" loading="lazy" />`
        : ""}
    </div>`;
}

function chipLabel(p) {
  const st = p.status || "pending";
  const stars = p.rating ? ` · ${"★".repeat(p.rating)}${"☆".repeat(5 - p.rating)}` : "";
  return `P${p.index} · ${st}${stars}`;
}

function curationBlock(story) {
  const c = story.curation || {};
  const panels = (c.panels || []).slice().sort((a, b) => (a.index || 0) - (b.index || 0));
  const avg = c.avg_rating != null ? ` · avg ${c.avg_rating}/5` : "";
  const chips = panels
    .map((p) => {
      const st = p.status || "pending";
      const tip = [p.note, p.suggestions].filter(Boolean).join(" — ") || st;
      return `<button type="button" class="curation-chip status-${esc(st)}" data-story="${esc(story.id)}" data-panel="${esc(p.index)}" title="${esc(tip)}">${esc(chipLabel(p))}</button>`;
    })
    .join("");
  return `
    <div class="curation-block">
      <div class="curation-head">
        <h3>Curation · Phase 8.5</h3>
        <p class="muted small">
          episode: <strong>${esc(c.episode_status || "pending")}</strong>
          · approved ${esc(c.approved || 0)}
          · rejected ${esc(c.rejected || 0)}
          · pending ${esc(c.pending || 0)}
          · regen ${esc(c.regenerated || 0)}${esc(avg)}
        </p>
      </div>
      <div class="link-row">
        <button type="button" class="btn act-approve-ep" data-story="${esc(story.id)}">Approve episode</button>
        <button type="button" class="btn act-reject-ep" data-story="${esc(story.id)}">Reject episode</button>
      </div>
      <div class="curation-chips">${chips || '<span class="muted small">No panel rows — click Seed.</span>'}</div>
      <p class="muted small">Click chip = rate / edit prompt · Shift = reject editor · Alt = approve quickly.</p>
    </div>`;
}

async function patchCuration(storyId, body) {
  const res = await fetch(`/api/curation/${encodeURIComponent(storyId)}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

function openPanelEditor(storyId, panel, mode = "edit") {
  PanelEditor.open({
    storyId,
    panel,
    mode,
    onDone: async () => {
      await loadLibrary(true);
    },
  });
}

function bindCurationActions() {
  document.querySelectorAll(".act-approve-ep").forEach((btn) => {
    btn.addEventListener("click", async () => {
      await patchCuration(btn.dataset.story, { status: "approved", note: "episode approved in library" });
      await loadLibrary(true);
    });
  });
  document.querySelectorAll(".act-reject-ep").forEach((btn) => {
    btn.addEventListener("click", async () => {
      const note = prompt("Reject episode — why?", "needs work") || "rejected";
      await patchCuration(btn.dataset.story, { status: "rejected", note });
      await loadLibrary(true);
    });
  });
  document.querySelectorAll(".curation-chip").forEach((btn) => {
    btn.addEventListener("click", async (ev) => {
      const story = btn.dataset.story;
      const panel = Number(btn.dataset.panel);
      try {
        if (ev.altKey) {
          await patchCuration(story, { status: "approved", panel, note: "approved in library" });
          await loadLibrary(true);
        } else if (ev.shiftKey) {
          openPanelEditor(story, panel, "reject");
        } else {
          openPanelEditor(story, panel, "edit");
        }
      } catch (err) {
        alert(String(err.message || err));
        await loadLibrary(true);
      }
    });
  });
}

async function loadLibrary(refresh = false) {
  const url = refresh ? "/api/library?refresh=true" : "/api/library";
  const res = await fetch(url);
  const data = await res.json();
  $("updated").textContent = data.updated_at
    ? `updated ${new Date(data.updated_at).toLocaleString()}`
    : "ready";
  const cs = data.curation_summary || {};
  const by = (cs.by_status || []).map((x) => `${x.status}:${x.n}`).join(" · ");
  $("libMeta").textContent = `${data.count || 0} stories · store ${data.store_path || "outputs/library/catalog.json"}${by ? ` · curation ${by}` : ""}`;

  const stories = data.stories || [];
  if (!stories.length) {
    $("libraryGrid").innerHTML = `<p class="muted" style="padding:0 2rem">No stories in library yet.</p>`;
    return;
  }

  $("libraryGrid").innerHTML = stories
    .map((s) => {
      const eds = s.editions || {};
      return `
        <article class="library-card">
          <div class="library-head">
            <div>
              <p class="eyebrow">${esc(s.phase)} · ${esc(s.panel_count)} panels</p>
              <h2>${esc(s.title)}</h2>
              <p class="muted">${esc(s.topic)}</p>
            </div>
            <div class="link-row">
              <a class="btn" href="${esc(s.reader_href)}">Open reader</a>
              <a class="btn" href="${esc(s.json_href)}" target="_blank" rel="noreferrer">JSON</a>
            </div>
          </div>
          ${curationBlock(s)}
          <div class="edition-grid">
            ${editionBlock("Reader edition", eds.reader)}
            ${editionBlock("Only Image", eds.image_only)}
          </div>
        </article>`;
    })
    .join("");
  bindCurationActions();
}

$("refreshLib").addEventListener("click", async () => {
  $("updated").textContent = "refreshing…";
  await fetch("/api/library/refresh", { method: "POST" });
  await loadLibrary(true);
});

$("seedCurate").addEventListener("click", async () => {
  $("updated").textContent = "seeding…";
  await fetch("/api/curation/seed", { method: "POST" });
  await loadLibrary(true);
});

loadLibrary();
loadAuthChrome();
